pkg/util/limiter/per_query_limiter.go

```diff
@@ -0,0 +1,119 @@
+package limiter
+
+import (
+	"context"
+	"fmt"
+	"github.com/cortexproject/cortex/pkg/cortexpb"
+	"github.com/cortexproject/cortex/pkg/ingester/client"
+	"github.com/cortexproject/cortex/pkg/util"
+	"github.com/cortexproject/cortex/pkg/util/validation"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/pkg/labels"
+	"go.uber.org/atomic"
+	"sync"
+)
+
+type PerQueryLimiter struct {
+	uniqueSeriesMx sync.RWMutex
+	uniqueSeries   map[model.Fingerprint]struct{}
+
+	chunkBytesCount       *atomic.Int32
+	maxSeriesPerQuery     int
+	maxChunkBytesPerQuery int
+}
+type perQueryLimiterCtxMarker struct{}
```

Comment:
[nit] I would call it `...CtxKey` instead of `CtxMarker` given it's a context key.